lib/ansible/utils/ssh_agent.py

```diff
@@ -0,0 +1,666 @@
+                public_bytes = private_key.public_key().public_bytes(
+                    encoding=serialization.Encoding.Raw,
+                    format=serialization.PublicFormat.Raw,
+                )
+                private_bytes = private_key.private_bytes(
+                    encoding=serialization.Encoding.Raw,
+                    format=serialization.PrivateFormat.Raw,
+                    encryption_algorithm=serialization.NoEncryption()
+                )
+                return Ed25519PrivateKeyMsg(
+                    KeyAlgo.ED25519,
+                    binary_string(public_bytes),
+                    binary_string(private_bytes + public_bytes),
+                )
+            case _:
+                raise NotImplementedError(private_key)
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class RSAPrivateKeyMsg(PrivateKeyMsg):
+    type: KeyAlgo
+    n: mpint
+    e: mpint
+    d: mpint
+    iqmp: mpint
+    p: mpint
+    q: mpint
+    comments: unicode_string = dataclasses.field(default=unicode_string(''), compare=False)
+    constraints: constraints = dataclasses.field(default=constraints(b''))
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class DSAPrivateKeyMsg(PrivateKeyMsg):
+    type: KeyAlgo
+    p: mpint
+    q: mpint
+    g: mpint
+    y: mpint
+    x: mpint
+    comments: unicode_string = dataclasses.field(default=unicode_string(''), compare=False)
+    constraints: constraints = dataclasses.field(default=constraints(b''))
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class EcdsaPrivateKeyMsg(PrivateKeyMsg):
+    type: KeyAlgo
+    ecdsa_curve_name: unicode_string
+    Q: binary_string
+    d: mpint
+    comments: unicode_string = dataclasses.field(default=unicode_string(''), compare=False)
+    constraints: constraints = dataclasses.field(default=constraints(b''))
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class Ed25519PrivateKeyMsg(PrivateKeyMsg):
+    type: KeyAlgo
+    enc_a: binary_string
+    k_env_a: binary_string
+    comments: unicode_string = dataclasses.field(default=unicode_string(''), compare=False)
+    constraints: constraints = dataclasses.field(default=constraints(b''))
+
+
+@dataclasses.dataclass
+class PublicKeyMsg(Msg):
+    @staticmethod
+    def get_dataclass(
+            type: KeyAlgo
+    ) -> type[t.Union[
+            RSAPublicKeyMsg,
+            EcdsaPublicKeyMsg,
+            Ed25519PublicKeyMsg,
+            DSAPublicKeyMsg
+    ]]:
+        match type:
+            case KeyAlgo.RSA:
+                return RSAPublicKeyMsg
+            case KeyAlgo.ECDSA256 | KeyAlgo.ECDSA384 | KeyAlgo.ECDSA521:
+                return EcdsaPublicKeyMsg
+            case KeyAlgo.ED25519:
+                return Ed25519PublicKeyMsg
+            case KeyAlgo.DSA:
+                return DSAPublicKeyMsg
+            case _:
+                raise NotImplementedError(type)
+
+    def public_key(self) -> CryptoPublicKey:
+        type = self.type  # type: ignore[attr-defined]
```

Comment:
Why is this needed? It kinda defeats the purpose of typing if we don't actually declare arguments..
Do we need another intermediate dataclass? Or maybe a typed dict or smth?